### ISSUE TITLE
[feature requests] Scale, Repeat and shapes

### ISSUE THREAD

**Issue by thegwydd** [bjorn/tiled#2659]
Hi, I'd like to know if there's a way to allow a tile to use multiple cells using scale and repeat.

May be with an example it could be more clear:
Imagine super mario platform made of n yellow rectangles (let's say 5 for example). Instead of 5 objects it would be nice to have 1 object with a x5 scale on X and a x5 repet (uv mapping) on X.

This way it would be more easy and efficient to import data into engines that supports scale and repeat.

Another question is: 
Is it possible to assign a child group of shapes to an object? 
Referring to the previous example, it would be nice to assign a rectangle to the platform tile object and scale it with same values of the parent object.
This way, it would be a lot simpler to define collision areas of objects instead of having another level with all collision areas that are not related to the object.

**Comment by bjorn** [comment]
Scaling is possible, but repeating the tile image is not. It's definitely something that would be nice to add. However, there are a lot of ways in which objects could be made "smarter", so instead of hardcoding support for certain specific features it may be preferable to tie this into the support for scripting somehow.

On the other hand, just supporting a "repeatX/Y" mode on tile objects alongside the current "stretch" mode seems easy enough to add without other major changes.

You can assign the collision rectangle in the Tile Collision Editor, and it will scale along with the object (as least its visualization does this, but it's something you'd have to implement as such on your side as well).

**Comment by bjorn** [comment]
> On the other hand, just supporting a "repeatX/Y" mode on tile objects alongside the current "stretch" mode seems easy enough to add without other major changes.

(but of course, it's a bit more complicated than a stretch/repeat toggle, since you may want to repeat and scale at the same time...)

**Comment by thegwydd** [comment]
@bjorn I don't know if you agree but I think scale/repeat are common tools related to placement so they could be part of non-scripting related stuff.
I don't see such attributes not really "specific".

**Comment by bjorn** [comment]
> @bjorn I don't know if you agree but I think scale/repeat are common tools related to placement so they could be part of non-scripting related stuff.

Yeah, it could be added as built-in feature. First we'd need to define exactly how these properties are supposed to behave though.

Objects don't currently have a scale, they have a size. And the tile image is stretched to always match the size of its object. The tile image also has an offset, and when the size of the object does not match the size of the tile, this offset is scaled accordingly as well. Finally, the tile image is bottom-left aligned with the object origin / position, except in isometric maps, where it is bottom-center.

So if we repeat the tile image instead of stretching it, this could be expected to affect the way we handle the offset, since now it shouldn't scale. And what if the object is 1.5 times the tile image, where does half of the image get cut off? Is that based on the alignment? And would the additional "scale" property then just make the repeated image bigger, so that less repetitions fit on the object, or does it scale the object entirely?

Personally I'm thinking, that we have already over-complicated the object representation. A number of features have been added in various places that all come together and determine how the tile object appears on the map. It is affected by the tileset, the map and finally the object itself. If we add more stuff, it makes things more complicated due to the combination of stuff, and this may not only affect rendering but also editing (like in the case of the scaled drawing offset and when scaling the entire object).

So it's hard to move forward without making an even bigger mess. In which case it may be good to take a step back and consider what we really want to achieve.

Please let me know if you think I'm just over-complicating things in my head.

**Comment by thegwydd** [comment]
Of course you have a lot more knowledge about features than what I have so, for me, it's hard to tell you if you over-complicated things.
I don't know about Isometric maps (I didn't thought about them) as I was referring orthogonal maps only.

I have an idea that could make the feature easier to implement but it would require a change in the way you save tmx file.

The idea is based on "cell modifier" objects. 
Basically you already have all you need to draw a single tile in a cell. 
Imagine a wood box image, You load a 64x64 image but you want to use only bottom right quadrant of the image, containing the wood box.
You'll then have an offset of 32x32 and a size of 32x32.
That will define the portion of the image you'll use when you render a Tile in the map.

Now imagine the orthogonal map of 100x100 cells with 32,32 size.
At the cell 5,3 you place a wood box. It's a single tile, rendered as always.
By using (mouse|keyboard combo|anything) you move right and the tile doubles. The single tile is rendered twice. Cell 5,3 will refer to the same Tile but the modifier says Scale (2,1) and Repeat (2,1). 
Cell 6,3 is empty!
Now, with the same input (mouse|keyboard combo|anything) you move down. The single Tile is rendered now 4 times, because we fill a 64x64 space with a single object.
Cell 5,3 will refer to the same Tile but the modifier says Scale (2,2) and Repeat (2,2). 
Cell 6,3-5,4-6,4 are empty!

If we speak of an object we can do the same but here we need to make a distinction. The object may have a collision shape.
When defining the shape we should have a per-shape flag that says the shape can be stretched by Cell modifier.
This way we render the object as a single object but, as for the tile, we can span it over multiple cells and the collision shapes that have such flag enabled will be scaled (only scaled and not repeated) following Cell scale modifier.

Now we can create a floors with a single 32x32 floor tile that uses 50 horizontal cells and will have a single collision object that will use the entire stretched size.
1 Tile, 1 Collision object, a 50x1 cell!

To me sounds nice and quite quick for the user. Define an object, draw collision box, enable a flag, move to the map view, enable a couple of check boxes and start drawing with scale&repeat.

But now the problems. How can we save such cells modifiers? Instead of cells you save only indexes of tiles. That's nice and simple.
The only idea that I have in mind is to generate a <cell_modifiers> element that holds following properties:
  - id - attribute (to be used by cell as index of what to draw)
  - tile id - attribute
  - <modifier type="scale|repeat|...">
         <scale x="..." y="...">

This allows you to add more cell modifiers in the future.
